### PR TITLE
Document unstable `--extern` options

### PR DESCRIPTION
**Pull request by ChrisDenton**
These are needed for Cargo's `build-std` feature and for anyone who wanted to do a similar thing outside of Cargo.

### CODE REVIEW

**Comment by ChrisDenton** [comment]
cc @ehuss @jyn514 

I'll admit I don't really know what I'm doing but hopefully this is a start.

File: src/doc/unstable-book/src/compiler-flags/extern options.md
Comment (jyn514) [inline src/doc/unstable-book/src/compiler-flags/extern options.md:3]: Is this true for all the modifiers, or just noprelude?

```suggestion
This is unstable feature, so you have to provide `-Zunstable-options` to enable it.
```

File: src/doc/unstable-book/src/compiler-flags/extern options.md
Comment (jyn514) [inline src/doc/unstable-book/src/compiler-flags/extern options.md:19]: can we link to https://doc.rust-lang.org/stable/rustc/lints/listing/allowed-by-default.html#unused-crate-dependencies here?

File: src/doc/unstable-book/src/compiler-flags/extern options.md
Comment (jyn514) [inline src/doc/unstable-book/src/compiler-flags/extern options.md:18]: can we link to https://doc.rust-lang.org/stable/rustc/lints/listing/warn-by-default.html#exported-private-dependencies here?

**Comment by jyn514** [comment]
Looks like highfive got confused - I'll go ahead and

r? @ehuss

**Comment by jyn514** [comment]
cc @petrochenkov

File: src/doc/unstable-book/src/compiler-flags/extern options.md
Comment (ChrisDenton) [reply src/doc/unstable-book/src/compiler-flags/extern options.md]: Seems to be true of all modifiers:

```
Z:\test> rustc main.rs --extern priv:core=libcore.rlib
error: the `-Z unstable-options` flag must also be passed to enable `--extern options
```

File: src/doc/unstable-book/src/compiler-flags/extern options.md
Comment (jyn514) [inline src/doc/unstable-book/src/compiler-flags/extern options.md:19]: sorry, I was unclear - I'd rather not hard-code the stable channel in the docs. You can instead use relative links:
```suggestion
* `priv`: Mark the crate as a private dependency for the [`exported_private_dependencies`](https://doc.rust-lang.org/stable/rustc/lints/listing/warn-by-default.html#exported-private-dependencies) lint.
* `nounused`: Suppress [`unused-crate-dependencies`](../../rustc/lints/listing/allowed-by-default.html#unused-crate-dependencies) warnings for the crate.
```
at some point it would probably be nice to add templating for `{{channel}}`, like for rustdoc tests, but that doesn't need to block this PR.

Comment (ChrisDenton) [reply src/doc/unstable-book/src/compiler-flags/extern options.md:19]: Oh, sorry! I checked a few other pages in the unstable book that seemed to be using absolute links too.

Comment (jyn514) [reply src/doc/unstable-book/src/compiler-flags/extern options.md]: > I checked a few other pages in the unstable book that seemed to be using absolute links too.

yeah, it would be nice to fix those at some point ... especially for the unstable book these can change a lot between stable and nightly.

File: src/doc/unstable-book/src/compiler-flags/extern-options.md
Comment (ehuss) [inline src/doc/unstable-book/src/compiler-flags/extern-options.md:19]:
```suggestion
* `noprelude`: Do not add the crate to the external prelude. If used, it will need to be imported using `extern crate`.
  This is used by the [build-std project](https://github.com/rust-lang/wg-cargo-std-aware/) to simulate compatibility with sysroot-only crates.
```

**Review by ehuss — COMMENTED** [review]
Thanks!  Just a small addition.  These really should have tracking issues, but I never got around to it.

**Comment by ChrisDenton** [comment]
I should probably squish all these commits if that's ok?

**Comment by jyn514** [comment]
Yeah, I think it would be nice to squash :)

**Review by jyn514 — APPROVED** [review]

**Comment by ehuss** [comment]
Thanks!

@bors r+ rollup

**Comment by petrochenkov** [comment]
@ChrisDenton 
Could you make a tracking issue for this stuff?
(And also link to the item (3) in https://github.com/rust-lang/rust/pull/93901#issuecomment-1041325522 from it.)

**Comment by ChrisDenton** [comment]
I could but I'd prefer to defer to @ehuss or someone else if they have the time. I don't really know the implementation details or history. But I can do a best effort attempt if it would help get the ball rolling.

**Comment by ehuss** [comment]
I'll set up the tracking issues later today.

**Comment by ehuss** [comment]
Posted #98401 to add links to the tracking issues.

**Pull request merged**
